### PR TITLE
Sk/hist migration failure models

### PR DESCRIPTION
**Pull request by gsa-suk**
https://github.com/GSA-TTS/FAC/issues/2815
https://github.com/GSA-TTS/FAC/issues/2816

Screenshot 2023-11-22 at 12 43 00 PM
Screenshot 2023-11-22 at 12 43 36 PM

## PR checklist: submitters

- [ ]   Link to an issue if possible. If there’s no issue, describe what your branch does. Even if there is an issue, a brief description in the PR is still useful.
- [ ]   List any special steps reviewers have to follow to test the PR. For example, adding a local environment variable, creating a local test file, etc.
- [ ]   For extra credit, submit a screen recording like [this one](https://github.com/GSA-TTS/FAC/pull/1821).
- [ ]   Make sure you’ve merged `main` into your branch shortly before creating the PR. (You should also be merging `main` into your branch regularly during development.)
- [ ]   Make sure you’ve accounted for any migrations. When you’re about to create the PR, bring up the application locally and then run `git status | grep migrations`. If there are any results, you probably need to add them to the branch for the PR. Your PR should have only **one** new migration file for each of the component apps, except in rare circumstances; you may need to delete some and re-run `python manage.py makemigrations` to reduce the number to one. (Also, unless in exceptional circumstances, your PR should not delete any migration files.)
- [ ]   Make sure that whatever feature you’re adding has tests that cover the feature. This includes test coverage to make sure that the previous workflow still works, if applicable.
- [ ]   Make sure the full-submission.cy.js [Cypress test](https://github.com/GSA-TTS/FAC/blob/main/docs/testing.md#end-to-end-testing) passes, if applicable.
- [ ]   Do manual testing locally. Our tests are not good enough yet to allow us to skip this step. If that’s not applicable for some reason, check this box.
- [ ]   Verify that no Git surgery was necessary, or, if it was necessary at any point, repeat the testing after it’s finished.
- [ ]   Once a PR is merged, keep an eye on it until it’s deployed to dev, and do enough testing on dev to verify that it deployed successfully, the feature works as expected, and the happy path for the broad feature area (such as submission) still works.

## PR checklist: reviewers

- [ ]   Pull the branch to your local environment and run `make docker-clean; make docker-first-run && docker compose up`; then run `docker compose exec web /bin/bash -c "python manage.py test"`
- [ ]   Manually test out the changes locally, or check this box to verify that it wasn’t applicable in this case.
- [ ]   Check that the PR has appropriate tests. Look out for changes in HTML/JS/JSON Schema logic that may need to be captured in Python tests even though the logic isn’t in Python.
- [ ]   Verify that no Git surgery is necessary at any point (such as during a merge party), or, if it was, repeat the testing after it’s finished.

The larger the PR, the stricter we should be about these points.

### CODE REVIEW

**Comment by gsa-suk** [comment]
Updated model names and field names to match GSAFAC, removed routing to various databases.

**Review by sambodeme — CHANGES_REQUESTED** [review]
Let's create an ADR to document our choices and thought processes here. This could serve as a basis for technical discussion meeting.

File: backend/census_historical_migration/migrations/0002_changerecords_failedsacs.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+                ),
+                ("dbkey", models.TextField(blank=True, db_column="DBKEY", null=True)),
+                (
+                    "audityear",
+                    models.TextField(blank=True, db_column="AUDITYEAR", null=True),
+                ),
+                ("ein", models.TextField(blank=True, db_column="EIN", null=True)),
+                ("uei", models.TextField(blank=True, db_column="UEI", null=True)),
+                (
+                    "census_table_NAME",
```
Comment (phildominguez-gsa) [inline backend/census_historical_migration/migrations/0002_changerecords_failedsacs.py:36]: Should NAME be lowercase?

**Review by phildominguez-gsa — CHANGES_REQUESTED** [review]

**Comment by sambodeme** [comment]
@gsa-suk we can use this [ADR ](https://github.com/GSA-TTS/FAC/issues/2848) for discussion

File: backend/dissemination/models.py [hunk 2]
```diff
@@ -505,3 +506,19 @@ class SecondaryAuditor(models.Model):
     report_id = models.TextField(
         REPORT_ID_FK_HELP_TEXT,
     )
+
+
+class MigrationChangeRecord(models.Model):
+    audit_year = models.TextField(blank=True, null=True)
+    dbkey = models.TextField(blank=True, null=True)
+    sac = models.ForeignKey(
```
Comment (sambodeme) [inline backend/dissemination/models.py:514]: As discussed during the dev huddle, it seems that there is no consensus yet on what to do with the sac objects generated by historic data migration (we will most likely figure this very soon). In this event, the agreement is to delete these sac objects or move them elsewhere (like into the historic data database), we will need to change `sac` to `report_id` here and remove the ForeignKey. TBD

File: backend/census_historical_migration/admin.py [hunk 3]
```diff
@@ -1,3 +1,11 @@
 from django.contrib import admin  # noqa: F401
 
-# Register your models here.
+from census_historical_migration.models import (
+    ELECAUDITHEADER,
```
Comment (sambodeme) [inline backend/census_historical_migration/admin.py:4]: Any reason why `ELECAUDITHEADER` is part of this PR ?

File: backend/census_historical_migration/test_models.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
+from django.test import TestCase
+
+from model_bakery import baker
+
+from .models import ELECAUDITHEADER, ReportMigrationStatus, MigrationErrorDetail
+
+
+class CensusHistoricalMigrationTests(TestCase):
+    def test_can_load_elecauditheader_model(self):
```
Comment (sambodeme) [inline backend/census_historical_migration/test_models.py:9]: Same as my previous comment

File: backend/census_historical_migration/admin.py (hunk 3 above)
Comment (gsa-suk) [reply backend/census_historical_migration/admin.py:4]: This is inherited from the c2g branch.  Tests for other tables need to be added in future.

File: backend/dissemination/models.py (hunk 2 above)
Comment (gsa-suk) [reply backend/dissemination/models.py:514]: Based on Matt's note in the slack channel, sac will be replaced with record_id.

File: backend/census_historical_migration/admin.py [hunk 5]
```diff
@@ -1,3 +1,11 @@
 from django.contrib import admin  # noqa: F401
 
-# Register your models here.
+from census_historical_migration.models import (
+    ELECAUDITHEADER,
+    ReportMigrationStatus,
+    MigrationErrorDetail,
+)
+
+admin.site.register(ELECAUDITHEADER)
```
Comment (sambodeme) [inline backend/census_historical_migration/admin.py:9]: Let's add the remaining models then (`ELECEINS, ELECAUDITFINDINGS,..., ELEC***`). It's not necessary to include the test for all of them in this PR.

**Review by sambodeme — APPROVED** [review]
LGTM

**Review by phildominguez-gsa — APPROVED** [review]

**Review by JeanMarie-TTS — APPROVED** [review]

**Pull request merged**